### PR TITLE
Limit complexity of IntervalQueryBuilderTests#testRandomSource()

### LINKED ISSUE #41402: IntervalQueryBuilderTests#testMustRewrite failure
Test fails in master and can be reproduce with:

```
./gradlew :server:test --tests "org.elasticsearch.index.query.IntervalQueryBuilderTests.testMustRewrite"   -Dtests.seed=CD8F5E1129373035   -Dtests.security.manager=true   -Dtests.locale=os-RU   -Dtests.timezone=America/Scoresbysund   -Dcompiler.java=12   -Druntime.java=12
```
Stack trace:

```
 2> java.lang.IllegalArgumentException: Too many disjunctions to expand
  2> NOTE: leaving temporary files on disk at: /Users/ivera/projects/elasticsearch/server/build/testrun/test/temp/org.elasticsearch.index.query.IntervalQueryBuilderTests_CD8F5E1129373035-002
  2> NOTE: test params are: codec=Asserting(Lucene80), sim=Asserting(org.apache.lucene.search.similarities.AssertingSimilarity@b98403c), locale=os-RU, timezone=America/Scoresbysund
  2> NOTE: Mac OS X 10.14.3 x86_64/Oracle Corporation 12 (64-bit)/cpus=12,threads=1,free=431765552,total=536870912
  2> NOTE: All tests run in this JVM: [IntervalQueryBuilderTests]
```

Does not reproduce in 7.x so it is probably a side effect of updating the master to the last Lucene snapshot

### PR DESCRIPTION
**Pull request by romseygeek**
IntervalsSources can throw IllegalArgumentExceptions if they would produce
too many disjunctions.  To mitigate against this when building random
sources, we try and build a query up-front, and repeat with a different
random source if it is too complex.

Fixes #41402

### CODE REVIEW

**Comment by romseygeek** [comment]
This bug is only manifesting in the master branch at the moment, as the relevant changes in the lucene snapshot have not yet been backported to 7x, but we should backport this anyway to pre-empt failures when 7x upgrades its lucene version

**Review by iverase — APPROVED** [review]
LGTM.

Just a question of ignorance, what is the limit for number of disjunctions and realistically can the user hit it?

**Comment by romseygeek** [comment]
> what is the limit for number of disjunctions

`BooleanQuery.getMaxClauseCount()` - so it's not out of the question that you'd hit it, but it should be pretty unlikely, and it would be something we'd recommend against doing anyway.

File: server/src/test/java/org/elasticsearch/index/query/IntervalQueryBuilderTests.java [hunk 1]
```diff
@@ -49,7 +49,18 @@
 
     @Override
     protected IntervalQueryBuilder doCreateTestQueryBuilder() {
-        return new IntervalQueryBuilder(STRING_FIELD_NAME, createRandomSource());
+        while (true) {
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/index/query/IntervalQueryBuilderTests.java]: Can we instead limit the number the number of recursion that we allow in `createRandomSource()` ? More than 2-3 levels seem overkill to test.

**Review by jimczi — COMMENTED** [review]
I left one comment

File: server/src/test/java/org/elasticsearch/index/query/IntervalQueryBuilderTests.java (hunk 1 above)
Comment (romseygeek) [reply server/src/test/java/org/elasticsearch/index/query/IntervalQueryBuilderTests.java]: I pushed 21d9f121be19a48d525bb7e7f576e068409f7766

**Review by jimczi — APPROVED** [review]
LGTM thanks

**Pull request merged**